packages/packages_test.go

```diff
@@ -0,0 +1,388 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package packages
+
+import (
+	"context"
+	"testing"
+
+	"github.com/Masterminds/semver/v3"
+)
+
+func TestPackagesFilter(t *testing.T) {
+	filterTestPackages := []filterTestPackage{
+		{
+			Name:          "apache",
+			Version:       "1.0.0-rc1",
+			Release:       "beta",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "apache",
+			Version:       "1.0.0",
+			Release:       "ga",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "apache",
+			Version:       "2.0.0-rc2",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "nginx",
+			Version:       "1.0.0",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "nginx",
+			Version:       "2.0.0",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "mysql",
+			Version:       "0.9.0",
+			Release:       "experimental",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "logstash",
+			Version:       "1.1.0",
+			Release:       "experimental",
+			Type:          "integration",
+			KibanaVersion: "^7.17.0 || ^8.0.0",
+		},
+		{
+			Name:          "etcd",
+			Version:       "1.0.0-rc1",
+			Type:          "integration",
+			KibanaVersion: "^8.0.0",
+		},
+		{
+			Name:          "etcd",
+			Version:       "1.0.0-rc2",
+			Type:          "integration",
+			KibanaVersion: "^8.0.0",
+		},
+	}
+	packages := buildFilterTestPackages(filterTestPackages)
+
+	cases := []struct {
+		Title    string
+		Filter   Filter
+		Expected []filterTestPackage
+	}{
+		{
+			Title: "not matching package name",
+			Filter: Filter{
+				PackageName: "unknown",
+			},
+			Expected: []filterTestPackage{},
+		},
+		{
+			Title: "not matching package version",
+			Filter: Filter{
+				PackageName:    "apache",
+				PackageVersion: "1.2.3",
+			},
+			Expected: []filterTestPackage{},
+		},
+		{
+			Title: "prerelease package with experimental release flag default search",
+			Filter: Filter{
+				PackageName: "mysql",
+			},
+			Expected: []filterTestPackage{},
+		},
+		{
+			Title: "prerelease package with experimental release flag prerelease search",
+			Filter: Filter{
+				PackageName: "mysql",
+				Prerelease:  true,
+			},
+			Expected: []filterTestPackage{
+				// FIXME: This package should be returned.
+				// {Name: "mysql", Version: "0.9.0"},
+			},
+		},
```

Comment:
Yes, this is what I mention in the FIXMEs, and what would be fixed by the changes in https://github.com/elastic/package-registry/pull/891 around https://github.com/elastic/package-registry/pull/891/files#diff-20fd7b81f7ff43064db94851e0fc5ad6a9852949b07d72a02d48751be912534fL315

The problem is not with 0.x.y, versions, it is that these 0.x.y packages had `release: experimental`, and they are discarded by the condition checking experimental packages.

During the changes around the prerelease label (https://github.com/elastic/package-spec/issues/225), there was also a push to move packages that we want to maintain out of experimental. This is why you can find multiple examples of 0.x packages that were experimental, updated directly to non-experimental 1.x versions. Most of the packages that only have versions with `release: experimental` are packages that we don't want to maintain on its current form, or have been replaced by other packages.

There was also some back and forth about how `prerelease` flag should behave on regard to the `release` flag (and how `experimental` should behave with semantic versioning), I don't remember the details, but we ended up with the implementation in https://github.com/elastic/package-registry/pull/785.

Now, looking to the unit tests, there are some cases like the ones you mention that look wrong. But I am on the fence about "fixing" this, because this is also an opportunity to review packages that still use `release: experimental`. If they are to be maintained, new versions should be released following current approach, otherwise they won't be available after the change in Kibana (https://github.com/elastic/kibana/issues/122973) is done.

In this PR I only wanted to add the unit tests, to make more clear what would change if we apply more changes of https://github.com/elastic/package-registry/pull/891 in other PRs.